Astro 5.6 blog (#1494)
withastro/astro.build: src/content/blog/astro-560.mdx

```js
import { defineMiddleware } from 'astro:middleware';
import { API_URL } from 'astro:env/server';
import { createClient } from './client.js';

// Astro 5.5: undefined
// Astro 5.6: string
const client = createClient(API_URL);

export const onRequest = defineMiddleware((ctx, next) => {
	ctx.locals.client = client;
	return next();
});
```

> This improvement creates a more consistent developer experience across all Astro server environments.

Reviewer: non-blocking nit: Feels like this ideally wants a second sentence here for flow. Could even be light and humourous!

## Experimental sessions on Cloudflare

> The [Astro sessions API](https://docs.astro.build/en/reference/experimental-flags/sessions/) is an experimental feature that allows you to easily store user data between requests. It uses pluggable backends for storage, which are automatically configured when using the Node or Netlify. With Astro 5.6, we've brought that simplified integration to Cloudflare too. Astro now automatically configures Cloudflare KV storage when you're using sessions with the Cloudflare adapter. This means your session data can be reliably stored and accessed across Cloudflare's global network with minimal configuration.

Reviewer: ```suggestion
The [Astro sessions API](https://docs.astro.build/en/reference/experimental-flags/sessions/) is an experimental feature that allows you to easily store user data between requests. It uses pluggable backends for storage, which are automatically configured when using the Node or Netlify adapters.

With Astro 5.6, we've brought that simplified integration to Cloudflare, too. Astro now automatically configures Cloudflare KV storage when you're using sessions with the Cloudflare adapter. This means your session data can be reliably stored and accessed across Cloudflare's global network with minimal configuration.
```

## New Prefetch Eagerness Option

Find the perfect balance between speed and resource usage with Astro's new prefetch eagerness controls. With the experimental [`clientPrerender`](https://docs.astro.build/en/reference/experimental-flags/client-prerender/) flag enabled, you can use the `eagerness` option on `prefetch()` to suggest to the browser how eagerly it should prefetch/prerender link targets.

This new option, which aligns with the browser's [Speculation Rules API](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type/speculationrules), gives you fine-grained control over how aggressively the browser should prefetch or prerender your links:

```astro
---
---
<script>
// Control prefetching eagerness
import { prefetch } from 'astro:prefetch';

// Let's be strategic about this resource-intensive page
prefetch('/data-heavy-dashboard', { eagerness: 'conservative' });

// This page is critical to the user journey, load it ASAP!
prefetch('/product-details', { eagerness: 'immediate' });

// For most pages, a balanced approach works best
prefetch('/about', { eagerness: 'moderate' });
</script>
```

> You can choose from three eagerness levels:

Reviewer: The spec has four levels, it also includes an `eager`. Looks like our implementation does, too? https://github.com/Marocco2/astro/blob/01ee0d047e782b40f8ef691324c23b5b353f2eae/packages/astro/src/prefetch/index.ts#L210

## Bug fixes

As always, we've been working hard on fixing issues since the [5.5 release](/blog/astro-550). See [the changelog](https://github.com/withastro/astro/blob/main/packages/astro/CHANGELOG.md) for all the details.

## Community

The Astro core team is [Ben Holmes](https://github.com/bholmesdev), [Caleb Jasik](https://github.com/jasikpark), [Chris Swithinbank](https://github.com/delucis), [Emanuele Stoppa](https://github.com/ematipico), [Erika](https://github.com/Princesseuh), [Florian Lefebvre](https://github.com/florian-lefebvre), [Fuzzy](https://github.com/aFuzzyBear), [HiDeoo](https://github.com/hideoo), [Luiz Ferraz](https://github.com/Fryuni), [Matt Kane](https://github.com/ascorbic), [Matthew Phillips](https://github.com/matthewp), [Nate Moore](https://github.com/natemoo-re), [Reuben Tier](https://github.com/TheOtterlord), [Sarah Rainsberger](https://github.com/sarah11918), and [Yan Thomas](https://github.com/yanthomasdev).

> Thanks to all the contributors who helped make Astro 5.6 possible: [Chris Swithinbank](https://github.com/delucis), [Edward Brunetiere](https://github.com/P4tt4te), [Emanuele Stoppa](https://github.com/ematipico), [Florian Lefebvre](https://github.com/florian-lefebvre), [Martin Trapp](https://github.com/martrapp), [Matt Kane](https://github.com/ascorbic), [Matthew Phillips](https://github.com/matthewp), [Reuben Tier](https://github.com/TheOtterlord), [Vardhaman Bhandari](https://github.com/Vardhaman619), [Marocco2](https://github.com/Marocco2) and [Yury Michurin](https://github.com/yurynix).

Reviewer: Do I not have any commits in any of these PRs? 😅

---
title: 'Astro 5.6'
description: 'Astro 5.6 brings first class astro:env and experimental session support to Cloudflare, and gives more control over prefetching.'
homepageLink:
  title: 'Astro 5.6'
  subtitle: 'Available now!'
publishDate: '2025-04-03'
authors:
  - matt
coverImage: '/src/content/blog/_images/astro-560/blog-post-5-6.webp'
socialImage: '/src/content/blog/_images/astro-560/og-astro-5-6.webp'
lang: 'en'
---

import BlogContentImage from '/src/components/BlogContentImage.astro';
import Mention from '/src/components/Mention.astro';

> **Astro 5.6 brings first class astro:env and experimental session support to Cloudflare, and gives more control over prefetching.**

Reviewer 2: ```suggestion
**Astro 5.6 brings first-class astro:env and experimental session support to Cloudflare, and gives more control over prefetching.**
```

☁️ Soar through the clouds with these new features in Astro:

- [**Global `astro:env` on Cloudflare**](#global-astroenv-on-cloudflare)
- [**Experimental sessions on Cloudflare**](#experimental-sessions-on-cloudflare)
- [**New prefetch eagerness option**](#new-prefetch-eagerness-option)
- [**Improved config validation**](#improved-config-validation)

To upgrade an existing project, use the automated `@astrojs/upgrade` CLI tool. Alternatively, upgrade manually by running the upgrade command for your package manager:

```sh
# Recommended:
npx @astrojs/upgrade

# Manual:
npm install astro@latest
pnpm upgrade astro --latest
yarn upgrade astro --latest
```

## Global `astro:env` on Cloudflare

> Since Astro 5 you have been able to access [type-safe environment variables using `astro:env`](https://docs.astro.build/en/guides/environment-variables/#type-safe-environment-variables). However, environment variables in Cloudflare were only accessible within a request — limiting when and where you could use them. For example, if you wanted to use a shared API client instance you couldn't use environment variables if instantiating outside of a request. This could be confusing, and was different from our other adapters.

Reviewer 2: comma nits
```suggestion
Since Astro 5 you have been able to access [type-safe environment variables using `astro:env`](https://docs.astro.build/en/guides/environment-variables/#type-safe-environment-variables). However, environment variables in Cloudflare were only accessible within a request — limiting when and where you could use them. For example, if you wanted to use a shared API client instance, you couldn't use environment variables if instantiating outside of a request. This could be confusing and was different from our other adapters.
```

```astro
---
export const prerender = false;
const cart = await Astro.session.get('cart');
---

<a href="/checkout">🛒 {cart?.length ?? 0} items</a>
```

For more information, see the [experimental sessions docs](https://docs.astro.build/en/reference/experimental-flags/sessions/).

## New Prefetch Eagerness Option

Find the perfect balance between speed and resource usage with Astro's new prefetch eagerness controls. With the experimental [`clientPrerender`](https://docs.astro.build/en/reference/experimental-flags/client-prerender/) flag enabled, you can use the `eagerness` option on `prefetch()` to suggest to the browser how eagerly it should prefetch/prerender link targets.

This new option, which aligns with the browser's [Speculation Rules API](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type/speculationrules), gives you fine-grained control over how aggressively the browser should prefetch or prerender your links:

```astro
---
---
<script>
// Control prefetching eagerness
import { prefetch } from 'astro:prefetch';

// Let's be strategic about this resource-intensive page
prefetch('/data-heavy-dashboard', { eagerness: 'conservative' });

> // This page is critical to the user journey, load it ASAP!
> prefetch('/product-details', { eagerness: 'immediate' });

Reviewer: ```suggestion
// This page is critical to the user journey, load it ASAP!
prefetch('/product-details'); // defaults to `{ eagerness: 'immediate' }`
```
Wondering about showing it like this, maybe, since it's the default and no one needs to write this out in full?

```astro
---
---
<script>
// Control prefetching eagerness
import { prefetch } from 'astro:prefetch';

// Let's be strategic about this resource-intensive page
prefetch('/data-heavy-dashboard', { eagerness: 'conservative' });

// This page is critical to the user journey, load it ASAP!
prefetch('/product-details'); // defaults to `{ eagerness: 'immediate' }`

// For most pages, a balanced approach works best
prefetch('/about', { eagerness: 'moderate' });
</script>
```

You can choose from three eagerness levels:

- `'immediate'`: Prefetch immediately, if resource limits allow
- `'eager'`: Link is likely to be needed, so fetch at the earliest opportunity
- `'moderate'`: Let the browser decide when to prefetch
- `'conservative'`: Only prefetch when highly likely to be needed

This feature is particularly valuable when dealing with large numbers of links where you might otherwise run into [browser limits in place to guard against over-speculating](https://developer.chrome.com/blog/speculation-rules-improvements#chrome-limits).

Thanks to community member [Marocco2](https://github.com/Marocco2) for contributing this feature!

## Custom Fetch Option for Prerendered Error Pages

> When an on-demand rendered page needs to display an error, it may need to fetch the prerendered error page from elsewhere. Currently, this is done by making a request using the default `fetch` implementation, which may not be suitable for all use cases. For example, the page may be served from a location that can't be called recursively, or the page may be stored elsewhere. Astro 5.6 adds a new optional `prerenderedErrorPageFetch` option in the Adapter API to allow adapters to provide custom implementations for fetching prerendered error pages.

Reviewer: ```suggestion
When an on-demand rendered page needs to display an error, your adapter may need to fetch a prerendered error page from somewhere other than your server. Currently, this is done by making a request using the default `fetch` implementation, which may not be suitable for all use cases. For example, the page may be served from a location that can't be called recursively, or the page may be stored elsewhere.

Astro 5.6 adds a new optional `prerenderedErrorPageFetch` option in the Adapter API to allow adapters to provide custom implementations for fetching prerendered error pages.
```
Can we define "elsewhere" a little more precisely? Somewhere other than... (your server? correct as necessary!)

Also, highlighting again that this is an *adapter* thing, relevant to how they work.

## Removed features from experimental SVG

The [experimental SVG RFC](https://github.com/withastro/roadmap/pull/1035) has been under heavy discussion, and for that we thank you all for chiming in! As result of discussions, we decided to *temporarily* remove certain features. This doesn't mean that we won't add them again, but this gives us time to have separate discussions and reach out a good balance between DX and good defaults.

These items are no longer available and must be removed from your code:

- The `title` prop has been removed until we can settle on the correct balance between developer experience and accessibility. Please replace any `title` props on your components with `aria-label`:
  ```diff
  - <Logo title="My Company Logo" />
  + <Logo aria-label="My Company Logo" />
  ```
- Sprite mode has been temporarily removed while we consider a new implementation that addresses how this feature was being used in practice. This means that there are no longer multiple `mode` options, and all SVGs will be inline. All instances of `mode` must be removed from your project as you can no longer control a mode:
  ```diff
  - <Logo mode="inline" />

>   + <Logo />

Reviewer: ```suggestion
  + <Logo /> // Always inline
```

Astro 5.6 lifts this restriction, meaning you can now access your environment variables globally throughout your server code and brings the Cloudflare adapter in line with the other official adapters. This is possible thanks to [enhancements in Cloudflare Workers](https://developers.cloudflare.com/changelog/2025-03-17-importable-env/) that are now implemented in Astro.

```js
import { defineMiddleware } from 'astro:middleware';
import { API_URL } from 'astro:env/server';
import { createClient } from './client.js';

// Astro 5.5: undefined
// Astro 5.6: string
const client = createClient(API_URL);

export const onRequest = defineMiddleware((ctx, next) => {
	ctx.locals.client = client;
	return next();
});
```

> This improvement creates a more consistent developer experience across all Astro server environments, and means you don't need to waste headspace working out what counts as request scope or global scope. 

Reviewer: ```suggestion
This improvement creates a more consistent developer experience across all Astro server environments, and means you don't need to worry about figuring out what counts as request scope or global scope. 
```
Nice how this all worked out and gave us something to put here! Just "uncolonizing" the language a bit, maybe?

Reviewer: It means you sound very British! 😂

But, this doesn't mean they're gone for good! Removing them now at this early stage gives us time to have separate discussions and build features back based on your feedback, striking the right balance between DX, accessibility best practices, and sensible defaults.

These items are no longer available and must be removed from your code:

- The `title` prop has been removed until we can settle on the correct balance between developer experience and accessibility. Please replace any `title` props on your components with `aria-label`:
  ```diff
  - <Logo title="My Company Logo" />
  + <Logo aria-label="My Company Logo" />
  ```
- Sprite mode has been temporarily removed while we consider a new implementation that addresses how this feature was being used in practice. This means that there are no longer multiple `mode` options, and all SVGs will be inline. All instances of `mode` must be removed from your project as you can no longer control a mode:
  ```diff
  - <Logo mode="inline" />
  + <Logo /> // Always inline
  ```

>   ```diff

Reviewer: ```suggestion
  
  ```diff
```
The two code samples were getting squished together without anything between them, making it look like one big (weird) code snippet. This should help that!
